src/ai/defaultai.cc

```diff
@@ -4293,6 +4295,26 @@ bool DefaultAI::check_productionsites(uint32_t gametime) {
 
 	const Map& map = game().map();
 
+	// First we check if we must release an experienced worker
+	// iterate over all working positions of the actual productionsite
+	for (uint8_t i = 0; i <  site.site->descr().nr_working_positions(); i++) {
+		// get the pointer to the worker assigned to the actual position
+		const Worker* cw = site.site->working_positions()[i].worker;
+		if (cw) {  // a worker is assigned to the position
+			// get the descritpion index of the worker assigned on this position
+			DescriptionIndex current_worker = cw->descr().worker_index();
+			// if description indexes of assigned worker and normal worker differ
+			// (this means an experienced worker is assigned to the position)
+			// and we have none of the experienced workers on stock
+			if (current_worker != site.bo->positions.at(i) &&
```

Comment:
Dont we need to check if we have junion for this position as well? If no junior is available, will the senior come back, will not he? - Just idea for consideration...

Reply:
that is a good question. normally if no junior exists and no tools exist this should be the case. However he might run into another building with an older request. Potetnially this might lead to a deadlock for barbarians and frisians with their smithes if not enough hammers could be produced in time. I wanted to keep it simple though and therefore I'd wait for some results if this ever happens. But it could be implemented if you deem it neccessary

Reply:
Can this lead to endless loop of senior being kicked and returning and kicked again from a single site?

Reply:
if no proper tool and no junior available and no other senior in any other site yes this is possible. But thats why I only kicked it out if no experinced worker on stock. 
So do you think I should add a check if junior or tool is available?

Reply:
Well, it is unambiguous. Also the fact whether the current site is or is not fully staffed can make difference. I dont have clear opinion what is best solution here :( Also, I am not that much into micromanagement... you can discuss it in forum, let other players say their opinion.

Reply:
I am afraid they won't have one as this is not affecting their gameplay. it is just the AI kicking workers out normal players need to do this by themselves.

Reply:
I think we should try with this solution and if we see AI's behaving weird on this I'll implement the check. this shouldn't affect training in any case I believe.

Reply:
OK, after some thinking and providing that workers ends the current working cycle (thus no material is wasted) I agree with you and no fatal problem will happen